File: app/src/main/java/com/sysu/edu/academic/CourseSelectionFilterFragment.java
package com.sysu.edu.academic;

import static android.text.TextUtils.isEmpty;
import static com.sysu.edu.api.CommonUtil.toStringOrDefault;

import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import androidx.activity.OnBackPressedCallback;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;

import com.alibaba.fastjson2.JSONArray;
import com.alibaba.fastjson2.JSONObject;
import com.google.android.material.textfield.MaterialAutoCompleteTextView;
import com.google.android.material.transition.MaterialContainerTransform;
import com.sysu.edu.api.Params;
import com.sysu.edu.databinding.FragmentCourseFilterBinding;
import com.sysu.edu.model.JwxtModel;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.stream.IntStream;


public class CourseSelectionFilterFragment extends Fragment {
    
    HashMap<String, String> filterValue = new HashMap<>();
    HashMap<String, String> filterName = new HashMap<>();
    CourseSelectionViewModel vm;
    FragmentCourseFilterBinding binding;
    NavController navController;
    JwxtModel model;
    
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        if (binding == null) {
            model = new JwxtModel(requireContext());
            vm = new ViewModelProvider(requireActivity()).get(CourseSelectionViewModel.class);
            filterValue = vm.getFilterValue();
            filterName = vm.getFilterName();
            binding = FragmentCourseFilterBinding.inflate(inflater, container, false);
            binding.container.setColumnCount(new Params(this).getColumn());
            model.getMessage().observe(requireActivity(), message -> {
                JSONObject response = message.getSecond();
                int what = message.getFirst();
                if (response.getInteger("code").equals(200)) {
                    JSONArray data = response.getJSONArray("data");
                    if (data != null) {
                        ArrayList<String> items = new ArrayList<>();
                        ArrayList<String> itemCodes = new ArrayList<>();
                        items.add("");
                        itemCodes.add("");
                        data.forEach(a -> {
                            items.add(((JSONObject) a).getString(new String[]{"campusName", "dataName", "minorName", "dataName", "dataName"}[what]));
                            itemCodes.add(((JSONObject) a).getString(new String[]{"id", "dataNumber", "sectionNumber", "dataNumber", "dataNumber"}[what]));
                        });
                        MaterialAutoCompleteTextView textView = new MaterialAutoCompleteTextView[]{binding.campus, binding.days, binding.sections, binding.languages, binding.special}[what];
                        textView.setSimpleItems(items.toArray(new String[]{}));
                        textView.setOnItemClickListener((_, _, i, _) -> {
                            filterValue.put(new String[]{"campus", "day", "section", "language", "special"}[what], itemCodes.get(i));
                            filterName.put(new String[]{"campus", "day", "section", "language", "special"}[what], items.get(i));
                        });
                    }
                }
                model.nextAll();
            });
            IntStream.range(0, 5).forEach(this::getData);
            load();
            model.next();
        }
        requireActivity().getOnBackPressedDispatcher().addCallback(
                getViewLifecycleOwner(),
                new OnBackPressedCallback(true) {
                    @Override
                    public void handleOnBackPressed() {
                        submit();
                    }
                }
        );
        return binding.getRoot();
    }
    
    void reset() {
        filterValue.clear();
        filterName.clear();
        vm.setFilterName(filterName);
        vm.setFilterValue(filterValue);
        load();
    }
    
    void load() {
        filterName = vm.getFilterName();
        filterValue = vm.getFilterValue();
        binding.campus.setText(filterName.getOrDefault("campus", ""), false);
        binding.course.setText(filterName.getOrDefault("course", ""));
        binding.days.setText(filterName.get("day"), false);
        binding.sections.setText(filterName.get("section"), false);
        binding.languages.setText(filterName.get("language"), false);
        binding.special.setText(filterName.getOrDefault("special", ""), false);
        binding.school.setText(filterName.get("school"));
        binding.teacher.setText(filterName.get("teacher"));
    }
    
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        view.setTransitionName("miniapp");
        navController = Navigation.findNavController(view);
        binding.reset.setOnClickListener(_ -> reset());
        binding.submit.setOnClickListener(_ -> submit());
        MaterialContainerTransform transition = new MaterialContainerTransform();
        transition.setScrimColor(Color.TRANSPARENT);
        transition.setAllContainerColors(requireContext().getColor(com.google.android.material.R.color.design_default_color_surface));
        setSharedElementEnterTransition(transition);
        setSharedElementReturnTransition(transition);
    }
    
    
    void submit() {
        vm.setReturnData(parseFilter(getMap()));
        vm.setFilterName(filterName);
        vm.setFilterValue(filterValue);
        navController.navigateUp();
    }
    
    void getData(int i) {
        model.add(new String[]{"jwxt/base-info/campus/findCampusNamesBox",
                "jwxt/base-info/codedata/findcodedataNames?datableNumber=233",
                "jwxt/base-info/AcadyeartermSet/minorName?schoolYear=2025-1",
                "jwxt/base-info/codedata/findcodedataNames?datableNumber=204",
                "jwxt/base-info/codedata/findcodedataNames?datableNumber=387"}[i], i);
    }
    
    HashMap<String, String> getMap() {
        filterValue.put("course", getEditText(binding.course));
        filterValue.put("teacher", getEditText(binding.teacher));
        filterValue.put("school", getEditText(binding.school));
        filterName.put("course", getEditText(binding.course));
        filterName.put("teacher", getEditText(binding.teacher));
        filterName.put("school", getEditText(binding.school));
        return filterValue;
    }
    
    String getEditText(EditText editText) {
        return toStringOrDefault(editText.getText());
    }
    
    String parseFilter(HashMap<String, String> filter) {
        StringBuilder str = new StringBuilder();
        String[] keys = new String[]{"course", "campus", "day", "section", "school", "teacher", "language", "special"};
        String[] key = new String[]{"courseName", "studyCampusId", "week", "classTimes", "courseUnitNum", "teachingTeacherNum", "teachingLanguageCode", "specialClassCode"};
        for (int i = 0; i < keys.length; i++) {
            String v = filter.getOrDefault(keys[i], "");
            if (!isEmpty(v)) str.append(String.format(",\"%s\":\"%s\"", key[i], v));
        }
        return str.toString();
    }
}

File: app/src/main/java/com/sysu/edu/academic/CourseSelectionFilterFragment.kt
package com.sysu.edu.academic

import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.EditText
import androidx.activity.OnBackPressedCallback
import androidx.navigation.fragment.findNavController
import androidx.navigation.fragment.navArgs
import com.alibaba.fastjson2.JSONObject
import com.google.android.material.R
import com.google.android.material.transition.MaterialContainerTransform
import com.sysu.edu.BaseFragment
import com.sysu.edu.databinding.FragmentCourseFilterBinding
import com.sysu.edu.model.JwxtModel

class CourseSelectionFilterFragment : BaseFragment() {
	var filterValue: CourseFilterValueData? = null
	var filterName: CourseFilterNameData? = null
	
	//lateinit var vm: CourseSelectionViewModel
	lateinit var binding: FragmentCourseFilterBinding
	lateinit var model: JwxtModel
	override fun onDestroyView() {
		super.onDestroyView()
		model.dispose()
	}
	
	private val args: CourseSelectionFilterFragmentArgs by navArgs()
	override fun onCreateView(inflater: LayoutInflater,
	                          container: ViewGroup?,
	                          savedInstanceState: Bundle?): View {
		super.onCreateView(inflater, container, savedInstanceState)
		model = JwxtModel(requireContext())		//vm = ViewModelProvider(requireActivity())[CourseSelectionViewModel::class.java]
		binding = FragmentCourseFilterBinding.inflate(inflater, container, false)
		binding.container.setColumnCount(config.column)
		model.message.observe(viewLifecycleOwner) { (code, response) ->
			if (response.getInteger("code") == 200) {
				val data = response.getJSONArray("data")
				if (data != null) {
					val items = mutableListOf<String?>()
					val itemCodes = mutableListOf<String?>()
					items.add("")
					itemCodes.add("")
					data.forEach { a: Any? ->
						items.add((a as JSONObject).getString(arrayOf("campusName", "dataName", "minorName", "dataName", "dataName")[code]))
						itemCodes.add(a.getString(arrayOf("id", "dataNumber", "sectionNumber", "dataNumber", "dataNumber")[code]))
					}
					val textView = arrayOf(binding.campus, binding.days, binding.sections, binding.languages, binding.special)[code]
					textView.setSimpleItems(items.toTypedArray())
					textView.setOnItemClickListener { _: AdapterView<*>?, _: View?, i: Int, _: Long ->
						filterValue?.set(arrayOf("campus", "day", "section", "language", "special")[code], itemCodes[i])
						filterName?.set(arrayOf("campus", "day", "section", "language", "special")[code], items[i])
					}
				}
			}
			model.nextAll()
		}
		(0..<5).forEach { getData(it) }
		model.next()
		requireActivity().onBackPressedDispatcher.addCallback(viewLifecycleOwner, object :
			OnBackPressedCallback(true) {
			override fun handleOnBackPressed() {
				submit()
			}
		})
		args.courseSelectionNameFilter?.let {
			filterName = it
			load()
		}
		args.courseSelectionValueFilter?.let {
			filterValue = it
		}
		return binding.root
	}
	
	private fun load() {
		binding.campus.setText(filterName?.studyCampusId, false)
		binding.course.setText(filterName?.courseName)
		binding.days.setText(filterName?.week, false)
		binding.sections.setText(filterName?.classTimes, false)
		binding.languages.setText(filterName?.teachingLanguageCode, false)
		binding.special.setText(filterName?.specialClassCode, false)
		binding.school.setText(filterName?.courseUnitNum)
		binding.teacher.setText(filterName?.teachingTeacherNum)
	}
	
	fun reset() {
		filterValue = CourseFilterValueData()
		filterName = CourseFilterNameData()
		load()
	}
	
	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		super.onViewCreated(view, savedInstanceState)
		view.transitionName = "miniapp"
		binding.reset.setOnClickListener { reset() }
		binding.submit.setOnClickListener { submit() }
		val transition = MaterialContainerTransform()
		transition.scrimColor = Color.TRANSPARENT
		transition.setAllContainerColors(requireContext().getColor(R.color.design_default_color_surface))
		sharedElementEnterTransition = transition
		sharedElementReturnTransition = transition
	}
	
	fun submit() {
		map
		findNavController().previousBackStackEntry?.savedStateHandle?.apply {
			set("filter_name", filterName)
			set("filter_value", filterValue)
		}
		findNavController().popBackStack()
	}
	
	fun getData(i: Int) {
		model.add(arrayOf("jwxt/base-info/campus/findCampusNamesBox", "jwxt/base-info/codedata/findcodedataNames?datableNumber=233", "jwxt/base-info/AcadyeartermSet/minorName?schoolYear=2025-1", "jwxt/base-info/codedata/findcodedataNames?datableNumber=204", "jwxt/base-info/codedata/findcodedataNames?datableNumber=387")[i], i)
	}
	
	val map: Unit
		get() {
			filterValue?.courseName = getEditText(binding.course)
			filterValue?.teachingTeacherNum = getEditText(binding.teacher)
			filterValue?.courseUnitNum = getEditText(binding.school)
			filterName?.courseName = getEditText(binding.course)
			filterName?.teachingTeacherNum = getEditText(binding.teacher)
			filterName?.courseUnitNum = getEditText(binding.school)
		}
	
	fun getEditText(editText: EditText): String {
		return editText.text.toString().trim()
	}
}